pyro/ops/tensor_utils.py

```diff
@@ -393,9 +394,19 @@ def inverse_haar_transform(x):
     return x
 
 
-def cholesky(x):
+def safe_cholesky(x):
     if x.size(-1) == 1:
+        if CHOLESKY_JITTER:
```

Comment:
Do you intend to use `CHOLESKY_JITTER * sqrt(finfo(x.dtype).eps)` here?

Reply:
My intention was to scale by about `finfo(x.dtype).eps * x.max()` so that the jitter was just barely detectable by the largest matrix entry before Cholesky factorizing.  That way if we set `RELATIVE_CHOLESKY_JITTER = 1/2` then jitter will only affect matrix entries less than half the size of the max.  And it kindof makes sense to me that each additional bit of precision would mean we would need to add half as much jitter, thus jitter would be proportional to `finfo(x.dtype).eps`.  Mostly the proportional think helps us keep a constant `RELATIVE_CHOLESKY_JITTER` across float32 and float64.

What's your intuition behind the square here, is that to keep constant error post-Cholesky-factorization?

Reply:
I missed the `x_max` term in the above comment. Using square seems to be more consistent w.r.t. to cases `x.size(-1) > 1` - but I like your clamp by `tiny` better.

Re `x_max`: using global `max` makes sense, but I feel that it might be better to use `max` of rows instead, e.g. considering the diagonal matrix `[0.0001, 10000]`, the global jitter is large w.r.t. the first diagonal term.

Reply:
Nice idea!  I've switched to using a row-wise max.  This required increasing `CHOLESKY_RELATIVE_JITTER` from 1.0 to 4.0, but this way still seems better :+1:

Reply:
is this really preferred? this changes the eigenvalues and eigenvectors as opposed to jitter that is proportional to the identity (which only changes eigenvalues)

Reply:
@martinjankowiak that's a good point, I didn't know about eigenvalue preservation.  I'd be ok with either version.

One thing I like about @fehiepsi's solution is that users can on their side rotate the system before performing Gaussian ops, e.g. I'm approximately diagonalizing via QR
```python
evals, evecs = torch.linalg.eig(transition_matrix)
Q, R = torch.linalg.qr(evecs.real)
transition_matrix = Q.T @ transition_matrix @ Q
```
which shrinks my diagonal perturbations
```diff
- [5.45, 1.81, 0.86, 0.76] * eps * CHOLESKY_RELATIVE_JITTER
+ [4.91, 1.67, 0.21, 0.17] * eps * CHOLESKY_RELATIVE_JITTER
```